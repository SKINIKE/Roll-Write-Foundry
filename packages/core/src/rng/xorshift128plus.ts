import { SerializedRngState } from '../types';

export type SeedInput = number | bigint | string | [bigint, bigint] | SerializedRngState;

const UINT64_MASK = (1n << 64n) - 1n;

function toUint64(value: bigint): bigint {
  return value & UINT64_MASK;
}

function splitmix64(seed: bigint): () => bigint {
  let state = toUint64(seed);
  return () => {
    state = toUint64(state + 0x9e3779b97f4a7c15n);
    let z = state;
    z = toUint64((z ^ (z >> 30n)) * 0xbf58476d1ce4e5b9n);
    z = toUint64((z ^ (z >> 27n)) * 0x94d049bb133111ebn);
    return toUint64(z ^ (z >> 31n));
  };
}

function hashString(seed: string): bigint {
  let hash = 0xcbf29ce484222325n;
  for (let i = 0; i < seed.length; i += 1) {
    hash ^= BigInt(seed.charCodeAt(i));
    hash = toUint64(hash * 0x100000001b3n);
  }
  return hash;
}

function normalizeSeed(seed: SeedInput | undefined): [bigint, bigint] {
  if (typeof seed === 'number') {
    return normalizeSeed(BigInt(seed));
  }
  if (typeof seed === 'bigint') {
    const generator = splitmix64(seed);
    return [generator(), generator() || 1n];
  }
  if (typeof seed === 'string') {
    const hash = hashString(seed);
    const generator = splitmix64(hash);
    return [generator(), generator() || 1n];
  }
  if (Array.isArray(seed)) {
    let [s0, s1] = seed;
    if (typeof s0 === 'number') s0 = BigInt(s0);
    if (typeof s1 === 'number') s1 = BigInt(s1);
    s0 = toUint64(BigInt(s0));
    s1 = toUint64(BigInt(s1));
    if (s0 === 0n && s1 === 0n) {
      s1 = 1n;
    }
    return [s0, s1];
  }
  if (seed && 'algorithm' in seed) {
    if (seed.algorithm !== 'xorshift128+') {
      throw new Error(`Unsupported RNG algorithm: ${seed.algorithm}`);
    }
    const [s0, s1] = seed.state;
    let state0 = toUint64(BigInt(s0));
    let state1 = toUint64(BigInt(s1));
    if (state0 === 0n && state1 === 0n) {
      state1 = 1n;
    }
    return [state0, state1];
  }
  return normalizeSeed(1n);
}

function toHex(value: bigint): string {
  return `0x${value.toString(16).padStart(16, '0')}`;
}

export class Xorshift128Plus {
  private state0: bigint;

  private state1: bigint;

  constructor(seed?: SeedInput) {
    const [s0, s1] = normalizeSeed(seed);
    this.state0 = s0;
    this.state1 = s1;
    if (this.state0 === 0n && this.state1 === 0n) {
      this.state1 = 1n;
    }
  }

  static deserialize(serialized: SerializedRngState): Xorshift128Plus {
    return new Xorshift128Plus(serialized);
  }

  clone(): Xorshift128Plus {
    const clone = new Xorshift128Plus();
    clone.state0 = this.state0;
    clone.state1 = this.state1;
    return clone;
  }

  getState(): [bigint, bigint] {
    return [this.state0, this.state1];
  }

  serialize(): SerializedRngState {
    return {
      algorithm: 'xorshift128+',
      state: [toHex(this.state0), toHex(this.state1)],
    };
  }

  nextBigInt(): bigint {
    let s1 = this.state0;
    const s0 = this.state1;
    this.state0 = s0;
    s1 ^= (s1 << 23n) & UINT64_MASK;
    s1 ^= s1 >> 17n;
    s1 ^= s0;
    s1 ^= s0 >> 26n;
    this.state1 = toUint64(s1);
    const result = toUint64(this.state1 + s0);
    return result;
  }

  nextFloat(): number {
    const value = this.nextBigInt();
    const fraction = Number(value >> 11n) / 2 ** 53;
    return fraction;
  }

  nextInt(maxExclusive: number): number {
    if (!Number.isInteger(maxExclusive) || maxExclusive <= 0) {
      throw new RangeError('maxExclusive must be a positive integer');
    }
    const bound = BigInt(maxExclusive);
    const threshold = UINT64_MASK - (UINT64_MASK % bound);
    while (true) {
      const value = this.nextBigInt();
      if (value < threshold) {
        return Number(value % bound);
      }
    }
  }

  nextRange(minInclusive: number, maxExclusive: number): number {
    if (!Number.isInteger(minInclusive) || !Number.isInteger(maxExclusive)) {
      throw new RangeError('Range boundaries must be integers');
    }
    if (maxExclusive <= minInclusive) {
      throw new RangeError('maxExclusive must be greater than minInclusive');
    }
    const delta = maxExclusive - minInclusive;
    return minInclusive + this.nextInt(delta);
  }

  next(): number {
    return this.nextFloat();
  }
}
