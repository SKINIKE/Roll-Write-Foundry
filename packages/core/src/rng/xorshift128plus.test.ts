import { describe, expect, it } from 'vitest';
import { Xorshift128Plus } from './xorshift128plus';

const seeds = [42, 1337, 20250920] as const;

describe('Xorshift128Plus', () => {
  it('produces deterministic integer sequences for known seeds', () => {
    const sequences = seeds.reduce<Record<number, number[]>>((acc, seed) => {
      const rng = new Xorshift128Plus(seed);
      acc[seed] = Array.from({ length: 100 }, () => rng.nextInt(1_000));
      return acc;
    }, {});
    expect(sequences).toMatchSnapshot('xorshift-sequences');
  });

  it('round-trips through serialization', () => {
    const rng = new Xorshift128Plus(123);
    const first = rng.nextInt(1000);
    const serialized = rng.serialize();
    const restored = new Xorshift128Plus(serialized);
    expect(restored.nextInt(1000)).toBe(rng.nextInt(1000));
    restored.nextInt(1000);
    expect(restored.serialize().state).not.toEqual(serialized.state);
    expect(first).toBeTypeOf('number');
  });

  it('generates bounded ranges without bias', () => {
    const rng = new Xorshift128Plus('range-test');
    const counts = Array.from({ length: 6 }, () => 0);
    const iterations = 6_000;
    for (let i = 0; i < iterations; i += 1) {
      const value = rng.nextRange(0, 6);
      counts[value] += 1;
    }
    const average = iterations / 6;
    counts.forEach((count) => {
      expect(Math.abs(count - average)).toBeLessThan(average * 0.1);
    });
  });
});
