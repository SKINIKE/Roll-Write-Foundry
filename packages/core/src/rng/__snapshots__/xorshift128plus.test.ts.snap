// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Xorshift128Plus > produces deterministic integer sequences for known seeds > xorshift-sequences 1`] = `
{
  "1337": [
    967,
    996,
    746,
    133,
    749,
    229,
    732,
    215,
    188,
    759,
    963,
    598,
    797,
    175,
    824,
    97,
    789,
    218,
    971,
    879,
    914,
    35,
    616,
    857,
    830,
    29,
    965,
    222,
    339,
    627,
    431,
    803,
    517,
    755,
    447,
    946,
    238,
    636,
    68,
    692,
    72,
    516,
    970,
    571,
    714,
    322,
    572,
    677,
    687,
    645,
    202,
    224,
    349,
    852,
    644,
    224,
    482,
    213,
    63,
    647,
    516,
    797,
    745,
    232,
    19,
    664,
    341,
    941,
    924,
    504,
    785,
    730,
    160,
    36,
    844,
    592,
    630,
    297,
    994,
    392,
    149,
    466,
    86,
    448,
    37,
    85,
    11,
    502,
    301,
    68,
    393,
    881,
    838,
    441,
    333,
    867,
    530,
    611,
    528,
    244,
  ],
  "20250920": [
    926,
    460,
    89,
    22,
    565,
    590,
    186,
    204,
    780,
    43,
    181,
    838,
    396,
    686,
    858,
    42,
    996,
    713,
    93,
    114,
    944,
    646,
    158,
    128,
    220,
    72,
    392,
    621,
    450,
    69,
    147,
    779,
    217,
    803,
    447,
    582,
    57,
    565,
    558,
    892,
    444,
    250,
    585,
    712,
    155,
    172,
    17,
    362,
    497,
    833,
    226,
    255,
    645,
    924,
    834,
    950,
    706,
    351,
    367,
    145,
    433,
    878,
    59,
    4,
    915,
    403,
    411,
    273,
    792,
    901,
    574,
    737,
    828,
    36,
    534,
    330,
    774,
    973,
    518,
    403,
    704,
    212,
    389,
    657,
    447,
    318,
    41,
    875,
    70,
    644,
    696,
    709,
    858,
    991,
    904,
    803,
    419,
    209,
    992,
    44,
  ],
  "42": [
    378,
    875,
    270,
    240,
    691,
    800,
    136,
    294,
    930,
    837,
    391,
    908,
    55,
    458,
    693,
    14,
    361,
    626,
    480,
    4,
    585,
    158,
    240,
    156,
    448,
    882,
    225,
    457,
    950,
    913,
    890,
    231,
    354,
    939,
    210,
    536,
    565,
    60,
    761,
    831,
    763,
    363,
    187,
    529,
    495,
    844,
    93,
    882,
    168,
    793,
    37,
    450,
    658,
    657,
    499,
    766,
    881,
    394,
    559,
    881,
    809,
    791,
    914,
    327,
    595,
    209,
    238,
    587,
    569,
    904,
    910,
    830,
    914,
    831,
    185,
    671,
    425,
    946,
    689,
    660,
    952,
    495,
    625,
    808,
    612,
    473,
    413,
    952,
    561,
    138,
    876,
    646,
    113,
    53,
    803,
    550,
    610,
    884,
    630,
    462,
  ],
}
`;
